proto/analyze.proto

```diff
@@ -0,0 +1,189 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+syntax = "proto3";
+
+import "proto/conjunction.proto";
+import "proto/concept.proto";
+import "proto/error.proto";
+
+package typedb.protocol;
+
+message Analyze {
+
+  message Req {
+    string query = 1;
+  }
+
+  message Res {
+    oneof result {
+      Error err = 1;
+      AnalyzedQuery ok = 2;
+    }
+
+    message AnalyzedQuery {
+      QueryStructure structure = 1;
+      QueryAnnotations annotations = 2;
+
+      message QueryStructure {
+        PipelineStructure query = 1;
+        repeated FunctionStructure preamble = 2;
+
+        message FunctionStructure {
+          PipelineStructure body = 1;
+          repeated ConjunctionStructure.Variable arguments = 2;
+          oneof returns {
+            ReturnOpStream stream = 3;
+            ReturnOpSingle single = 4;
+            ReturnOpCheck check = 5;
+            ReturnOpReduce reduce = 6;
+          };
+
+          message ReturnOpStream {
+            repeated ConjunctionStructure.Variable variables = 1;
+          }
+          message ReturnOpSingle {
+            string selector = 1;
+            repeated ConjunctionStructure.Variable variables = 2;
+          }
+          message ReturnOpCheck {}
+          message ReturnOpReduce {
+            repeated Reducer reducers = 1;
+          }
+        }
+
+        message PipelineStructure {
+          repeated ConjunctionStructure conjunctions = 1;
+          repeated PipelineStage stages = 2;
+          map<uint32, VariableInfo> variable_info = 3;
+          repeated ConjunctionStructure.Variable outputs = 4;
+
+          message VariableInfo {
+            string name = 1;
+          }
+
+          message PipelineStage {
+            oneof stage {
+              Match match = 1;
+              Insert insert = 2;
+              Put put = 3;
+              Update update = 4;
+              Delete delete = 5;
+              Select select = 6;
+              Sort sort = 7;
+              Require require = 8;
+              Offset offset = 9;
+              Limit limit = 10;
+              Distinct distinct = 11;
+              Reduce reduce = 12;
+            }
+
+            // They're all the same structure. Should we just use one?
```

Comment:
I quite like them separate! In Rust itll be very easy to unpack them identically